cpp/include/copc-lib/copc/extents.hpp

```diff
@@ -0,0 +1,170 @@
+    }
+
+    std::vector<std::shared_ptr<CopcExtent>> ExtraBytes()
+    {
+        std::vector<std::shared_ptr<CopcExtent>> extra_bytes;
+        extra_bytes.reserve(extents_.size() - las::PointBaseNumberDimensions(point_format_id_) + 3);
+        extra_bytes.assign(extents_.begin() + las::PointBaseNumberDimensions(point_format_id_) - 3, extents_.end());
+        return extra_bytes;
+    }
+
+    // Setters for python bindings
+    void Intensity(std::shared_ptr<CopcExtent> intensity) { extents_[0] = intensity; }
+    void ReturnNumber(std::shared_ptr<CopcExtent> return_number) { extents_[1] = return_number; }
+    void NumberOfReturns(std::shared_ptr<CopcExtent> number_of_returns) { extents_[2] = number_of_returns; }
+    void ScannerChannel(std::shared_ptr<CopcExtent> scanner_channel) { extents_[3] = scanner_channel; }
+    void ScanDirectionFlag(std::shared_ptr<CopcExtent> scan_direction_flag) { extents_[4] = scan_direction_flag; }
+    void EdgeOfFlightLine(std::shared_ptr<CopcExtent> edge_of_flight_line) { extents_[5] = edge_of_flight_line; }
+    void Classification(std::shared_ptr<CopcExtent> classification) { extents_[6] = classification; }
+    void UserData(std::shared_ptr<CopcExtent> user_data) { extents_[7] = user_data; }
+    void ScanAngle(std::shared_ptr<CopcExtent> scan_angle) { extents_[8] = scan_angle; }
+    void PointSourceID(std::shared_ptr<CopcExtent> point_source_id) { extents_[9] = point_source_id; }
+    void GpsTime(std::shared_ptr<CopcExtent> gps_time) { extents_[10] = gps_time; }
+    void Red(std::shared_ptr<CopcExtent> red)
+    {
+        if (point_format_id_ > 6)
+            extents_[1] = red;
+        else
+            throw std::runtime_error("CopcExtents::Red: This point format does not have Red");
+    }
+    void Green(std::shared_ptr<CopcExtent> green)
+    {
+        if (point_format_id_ > 6)
+            extents_[12] = green;
+        else
+            throw std::runtime_error("CopcExtents::Green: This point format does not have Green");
+    }
+    void Blue(std::shared_ptr<CopcExtent> blue)
+    {
+        if (point_format_id_ > 6)
+            extents_[13] = blue;
+        else
+            throw std::runtime_error("CopcExtents::Blue: This point format does not have Blue");
+    }
+
+    void NIR(std::shared_ptr<CopcExtent> nir)
+    {
+        if (point_format_id_ == 8)
+            extents_[14] = nir;
+        else
+            throw std::runtime_error("CopcExtents::NIR: This point format does not have NIR");
+    }
+
+    void ExtraBytes(std::vector<std::shared_ptr<CopcExtent>> extra_bytes)
+    {
+        if (extra_bytes.size() != (extents_.size() - las::PointBaseNumberDimensions(point_format_id_) + 3))
+            throw std::runtime_error("CopcExtents::ExtraBytes: Vector of extra byte must be the right length.");
+        std::copy(extra_bytes.begin(), extra_bytes.end(),
+                  extents_.begin() + las::PointBaseNumberDimensions(point_format_id_) - 3);
```

Comment:
not sure I understand where the 3 comes from, the extra_bytes come from the back of the extents_ array right?